### PR TITLE
Support 3.13

### LINKED ISSUE #76: Bump support for LSP 3.13.0
3.13.0 was released on 9/11/2018.

### PR DESCRIPTION
**Pull request by kjeremy**
This should bump us up to 3.13. It's probably best to review each commit separately.

Supercedes #78 
Fixes #76

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -2337,6 +2687,8 @@ pub struct ApplyWorkspaceEditResponse {
     pub applied: bool,
 }
 
+pub type CodeActionKind = String;
```
[the remark this answers is not in the document]
Comment (kjeremy) [reply src/lib.rs]: I saw that but wasn't sure if those values were open ended or not (or how best to encode that?)

File: src/lib.rs [hunk 2]
```diff
@@ -319,15 +440,37 @@ pub struct WorkspaceEdit {
...
+     *
      * Whether a client supports versioned document edits is expressed via
-     * `WorkspaceClientCapabilities.workspaceEdit.documentChanges`.
+     * `workspace.workspaceEdit.documentChanges` client capability.
+     *
+     * If a client neither supports `documentChanges` nor `workspace.workspaceEdit.resourceOperations` then
+     * only plain `TextEdit`s using the `changes` property are supported.
      */
     #[serde(skip_serializing_if = "Option::is_none")]
     pub document_changes: Option<Vec<TextDocumentEdit>>,
```
[the remark this answers is not in the document]
Comment (kjeremy) [reply src/lib.rs]: good catch!

File: src/lib.rs [hunk 3]
```diff
@@ -319,15 +440,37 @@ pub struct WorkspaceEdit {
...
 
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(untagged)]
+pub enum DocumentChanges {
+    Edits(Vec<TextDocumentEdit>),
+    Operations(Vec<DocumentChangeOperation>),
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(untagged)]
```
[the remark this answers is not in the document]
Comment (kjeremy) [reply src/lib.rs]: How would you deal with the different options like `CreateFileOptions` etc.?

[the remark this answers is not in the document]
Comment (kjeremy) [reply src/lib.rs]: oh wait I think I see...

[the remark this answers is not in the document]
Comment (kjeremy) [reply src/lib.rs]: I'm not sure this combination will work. `#[serde(other)]` only works on unit variants.

File: src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kjeremy) [reply src/lib.rs]: If I'm reading it right the `Vec` needs to be removed too.

File: src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (kjeremy) [reply src/lib.rs]: I THINK inserting a shim gives us the correct results:

```rust
#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
#[serde(untagged)]
pub enum DocumentChanges {
    Edits(Vec<TextDocumentEdit>),
    Operations(Vec<DocumentChangeOperation>),
}

#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
#[serde(untagged, rename_all="lowercase" )]
pub enum DocumentChangeOperation {
    Op(ResourceOp), // <------- In an ideal world `ResourceOp` wouldn't be needed
    Edit(TextDocumentEdit),
}

#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
#[serde(tag = "kind", rename_all="lowercase" )]
pub enum ResourceOp {
    Create(CreateFile),
    Rename(RenameFile),
    Delete(DeleteFile),
}
```

This looks right according to my test except that the `TextDocumentEdit`'s fields are swapped so the string equality fails.

File: src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Marwes) [reply src/lib.rs]: Could just add a few `&str` constants to make the available.

File: src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Marwes) [reply src/lib.rs]: Yep, since `DocumentChanges ` contains a `Vec`

**Comment by kjeremy** [comment]
Okay I've addressed your comments.

**Comment by kjeremy** [comment]
Should be good now. With bf12571 we can now run against a language client build with the latest `vscode-language-client` 5.1.0

**Comment by Marwes** [comment]
Thanks again! Released as 0.51

**Pull request merged**